---
title: "OrthoPaiR Package Vignette"
author: "Your Name"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
%\VignetteIndexEntry{OrthoPaiR Package Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  ---
  
  ```{r setup, include=FALSE}
knitr::opts_chunk$set(
collapse = TRUE,
comment = "#>"
)
```

# Introduction

The `OrthoPaiR` package provides tools for conducting syntenic orthologous gene pairing between given genomes. This vignette demonstrates how to use the package functions to perform syntenic orthologous gene pairing and analyze the results.

# Installation of Prerequisite Packages

Before using the `OrthoPaiR` package, make sure you have the following prerequisite packages installed:

```{r}
install.packages("BiocManager")
BiocManager::install(c("GenomicRanges", "GenomeInfoDb", "rhdf5", "rtracklayer", "Biostrings", "S4Vectors"))
install.packages(c("ggplot2", "data.table"))
```

Ensure you also have Miniprot and SibeliaZ installed and executable. You can follow their installation guides on their respective websites.

# Quick start
You can go through the OrthoPaiR pipeline using the `orthopair()` function that is a 
wrapper functions to execute a series of the functions to go through the OrthoPaiR 
pipeline.
```{r}
object <- orthopair(query_genome = query_genome,
                    subject_genome = subject_genome,
                    query_gff = query_gff,
                    subject_gff = subject_gff,
                    query_cds = query_cds,
                    subject_cds = subject_cds,
                    query_prot = query_prot,
                    subject_prot = subject_prot,
                    hdf5_path = hdf5_path,
                    conda = "/home/ftom/miniforge/miniforge3/bin/conda",
                    miniprot_bin = "miniprot",
                    miniprot_condaenv = "miniprot",
                    n_threads = 30,
                    omit_chr = "chrUn|chrSy")
```
The orthopair() function output a OrthoPaiRDB object that contains a link to the hdf5 
file storing syntenic ortholog pairing results. Those results can be accessed by
the following functions.

```{r}
txwise_summary <- summaryOrthoPaiR(object = object, gene = FALSE)
genewise_summary <- summaryOrthoPaiR(object = object, gene = TRUE)
genewise_orthopair <- getOrthoPaiR(object = object, gene = TRUE)
orphan_genewise <- getOrphan(object = object, gene = TRUE)
```
See the [Output results] section in this vignette for more details of output resutls.

# Creating a OrthoPaiRDB Object

First, create a `OrthoPaiRDB` object using the `makeOrthoPaiRDB` function. 
This object will hold all necessary input files and parameters.

```{r create-OrthoPaiRdb}
library(OrthoPaiR)
```


```{r create-OrthoPaiRdb}
query_genome <- system.file("extdata", "query_genome.fa", package = "OrthoPaiR")
subject_genome <- system.file("extdata", "subject_genome.fa", package = "OrthoPaiR")
query_gff <- system.file("extdata", "query.gff", package = "OrthoPaiR")
subject_gff <- system.file("extdata", "subject.gff", package = "OrthoPaiR")
query_cds <- system.file("extdata", "query_cds.fa", package = "OrthoPaiR")
subject_cds <- system.file("extdata", "subject_cds.fa", package = "OrthoPaiR")
query_prot <- system.file("extdata", "query_prot.fa", package = "OrthoPaiR")
subject_prot <- system.file("extdata", "subject_prot.fa", package = "OrthoPaiR")
out_dir <- tempdir()
hdf5_path <- file.path(out_dir, "OrthoPaiR.h5")

object <- makeOrthoPaiRDB(query_genome = "inst/extdata/query_genome.fa",
                          subject_genome = "inst/extdata/subject_genome.fa",
                          query_gff = "inst/extdata/query.gff",
                          subject_gff = "inst/extdata/subject.gff",
                          query_cds = "inst/extdata/query_cds.fa",
                          subject_cds = "inst/extdata/subject_cds.fa",
                          query_prot = "inst/extdata/query_prot.fa",
                          subject_prot = "inst/extdata/subject_prot.fa",
                          hdf5_path = hdf5_path)
```
