---
title: "Synog Package Vignette"
author: "Your Name"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Synog Package Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

# Introduction

The `Synog` package provides tools for conducting syntenic orthologous gene pairing between given genomes. This vignette demonstrates how to use the package functions to perform syntenic orthologous gene pairing and analyze the results.

# Installation of Prerequisite Packages

Before using the `Synog` package, make sure you have the following prerequisite packages installed:

```{r}
install.packages("BiocManager")
BiocManager::install(c("GenomicRanges", "GenomeInfoDb", "rhdf5", "rtracklayer", "Biostrings", "S4Vectors"))
install.packages(c("ggplot2", "data.table"))
```

Ensure you also have Miniprot and SibeliaZ installed and executable. You can follow their installation guides on their respective websites.

# Quick start
You can go through the Synog pipeline using the `runSynog()` function that is a 
wrapper functions to execute a series of the functions to go through the Synog 
pipeline.
```{r}
object <- runSynog(query_genome = query_genome,
                   subject_genome = subject_genome,
                   query_gff = query_gff,
                   subject_gff = subject_gff,
                   query_cds = query_cds,
                   subject_cds = subject_cds,
                   query_prot = query_prot,
                   subject_prot = subject_prot,
                   hdf5_path = hdf5_path,
                   sibeliaz_out_dir = "./sibeliaz_out",
                   sibeliaz_bin = "sibeliaz",
                   maf2synteny_bin = "maf2synteny",
                   conda = "/home/ftom/miniforge/miniforge3/bin/conda",
                   sibeliaz_condaenv = "sibeliaz",
                   miniprot_bin = "miniprot",
                   miniprot_condaenv = "miniprot",
                   miniprot_out_dir = "./miniprot_out",
                   n_threads = 30,
                   omit_chr = "chrUn|chrSy")
```
The runSynog() function output a SynogDB object that contains a link to the hdf5 
file storing syntenic ortholog pairing results. Those results can be accessed by
the following functions.

```{r}
txwise_summary <- summarySynog(object = object, gene = FALSE)
genewise_summary <- summarySynog(object = object, gene = TRUE)
synog_genewise <- getSynog(object = object, gene = TRUE)
orphan_genewise <- getOrphan(object = object, gene = TRUE)
genewise_split_summary <- summarySynog(object = object, gene = TRUE, split = TRUE)
synog_txwise <- getSynog(object = object, gene = FALSE)
synog_genewise_split <- getSynog(object = object, gene = TRUE, split = TRUE)
orphan_genewise_split <- getOrphan(object = object, gene = TRUE, split = TRUE)
```
See the [Output results] section in this vignette for more details of output resutls.

# Creating a SynogDB Object

First, create a `SynogDB` object using the `makeSynogDB` function. 
This object will hold all necessary input files and parameters.

```{r create-synogdb}
library(Synog)
```


```{r create-synogdb}
query_genome <- system.file("extdata", "query_genome.fa", package = "Synog")
subject_genome <- system.file("extdata", "subject_genome.fa", package = "Synog")
query_gff <- system.file("extdata", "query.gff", package = "Synog")
subject_gff <- system.file("extdata", "subject.gff", package = "Synog")
query_cds <- system.file("extdata", "query_cds.fa", package = "Synog")
subject_cds <- system.file("extdata", "subject_cds.fa", package = "Synog")
query_prot <- system.file("extdata", "query_prot.fa", package = "Synog")
subject_prot <- system.file("extdata", "subject_prot.fa", package = "Synog")
out_dir <- tempdir()
hdf5_path <- file.path(out_dir, "synog.h5")

object <- makeSynogDB(query_genome = "inst/extdata/query_genome.fa",
                      subject_genome = "inst/extdata/subject_genome.fa",
                      query_gff = "inst/extdata/query.gff",
                      subject_gff = "inst/extdata/subject.gff",
                      query_cds = "inst/extdata/query_cds.fa",
                      subject_cds = "inst/extdata/subject_cds.fa",
                      query_prot = "inst/extdata/query_prot.fa",
                      subject_prot = "inst/extdata/subject_prot.fa",
                      hdf5_path = hdf5_path)

# object <- makeSynogDB(query_genome = query_genome,
#                       subject_genome = subject_genome,
#                       query_gff = query_gff,
#                       subject_gff = subject_gff,
#                       query_cds = query_cds,
#                       subject_cds = subject_cds,
#                       query_prot = query_prot,
#                       subject_prot = subject_prot,
#                       hdf5_path = hdf5_path)
```

# Running Sibeliaz

Run Sibeliaz to identify locally collinear blocks (LCBs) between the query and subject genomes.
Usually, installation of SibeliaZ using conda or conda wrappers may be the easiest way. If you installed SibeliaZ using conda or conda wrappers, you need to specify the name of the conda environment for SibeliaZ. If your conda binary is not executable from your R environment, please specify the path to the conda binary.

```{r run-sibeliaz}
runSibeliaZ(object = object,
            out_dir = file.path(out_dir, "sibeliaz_out"), 
            conda = "/home/ftom/miniforge/miniforge3/bin/conda",
            condaenv = "sibeliaz", 
            run_sibeliaz = TRUE)
```


```{r run-sibeliaz}
sibeliaLCB2DF(object = object)
lcbClassify(object = object)
getLCBpairs(object = object)
```

# Finding Ortholog Pairs Based on CDS RBH

Find reciprocal best hits (RBH) for CDS sequences.

```{r find-rbh}
rbh(object = object, n_threads = 30)
```

# Running LCB-based RBBH Filtering

Anchor orthologous genes using LCB-based RBBH filtering.

```{r anchor-ortho}
anchorOrtho(object = object)
```

# Synteny-based RBH Filtering

Filter orthologous genes based on synteny.

```{r synteny-ortho}
syntenyOrtho(object = object, 
             omit_chr = "chrUn|chrSy",
             pident = 90, 
             evalue = 1e-100, 
             qcovs = 50)
```

# Gene-wise Orthologous Pairing

Perform gene-wise orthologous pairing.

```{r gene-ortho}
geneOrtho(object = object)
```

# Summarizing Results

Generate summaries and export results.

```{r summarize-results}
txwise_summary <- summarySynog(object = object, gene = FALSE)
genewise_summary <- summarySynog(object = object, gene = TRUE)
synog_genewise <- getSynog(object = object, gene = TRUE)
orphan_genewise <- getOrphan(object = object, gene = TRUE)
```

# Mapping Proteins with Miniprot

Map proteins using Miniprot to further refine orthologous pairs.

```{r map-prot}
mapProt(object = object,
        conda = "/home/ftom/miniforge/miniforge3/bin/conda", condaenv = "miniprot",
        n_core = 20,
        out_prefix = file.path(out_dir, "miniprot_out"))
```

# Updating Files and Re-running Analyses

Update files and re-run analyses for refined results.

```{r update-files}
createFASTA(object = object)
object <- updateFiles(object = object)
rbh(object = object, n_threads = 30)
anchorOrtho(object = object)
syntenyOrtho(object = object, 
             omit_chr = "chrUn|chrSy",
             pident = 90, 
             evalue = 1e-100, 
             qcovs = 50)
geneOrtho(object = object)
```

# Summarizing Refined Results

Generate refined summaries and export results.

```{r summarize-refined-results}
txwise_summary <- summarySynog(object = object, gene = FALSE)
genewise_summary <- summarySynog(object = object, gene = TRUE)
synog_genewise <- getSynog(object = object, gene = TRUE)
orphan_genewise <- getOrphan(object = object, gene = TRUE)
```

# Splitting Genes

Split genes based on the orthologous pairs identified.

```{r split-genes}
splitGenes(object = object)
genewise_split_summary <- summarySynog(object = object, gene = TRUE, split = TRUE)
synog_txwise <- getSynog(object = object, gene = FALSE)
synog_genewise_split <- getSynog(object = object, gene = TRUE, split = TRUE)
orphan_genewise_split <- getOrphan(object = object, gene = TRUE, split = TRUE)
```
